pkg/virtual/framework/dynamic/apiserver/openapi.go

```diff
@@ -0,0 +1,331 @@
+	apisv1alpha1 "github.com/kcp-dev/kcp/sdk/apis/apis/v1alpha1"
+)
+
+const (
+	// DefaultServiceCacheSize is the default size of the OpenAPI service cache.
+	DefaultServiceCacheSize = 50
+)
+
+// openAPIHandler implements the OpenAPI v3 handler for virtual workspaces.
+//
+// This handler generates the OpenAPI v3 spec by:
+//
+//   - Creating specs based on webservices registered with the virtual workspace API server. These are only /api, /apis
+//     and /version, without any subroute (e.g. /apis/<group>)
+//   - Creating a webservice for every /apis/<group> route available by utilizing the API discovery, then generating
+//     OpenAPI v3 specs from these webservices
+//   - Converting every APIDefinition into a CustomResourceDefinition and generating OpenAPI v3 specs for
+//     /apis/<group>/<version> from these CRDs using the kube-openapi library
+//
+// This is different to the regular API servers, which are generating the OpenAPI v3 spec by purely using the
+// webservices. This is not an option here because the virtual workspace API server has only three webservices
+// registered (/api, /apis, and /version, without any subroute).
+type openAPIHandler struct {
+	apiSetRetriever    apidefinition.APIDefinitionSetGetter
+	goRestfulContainer *restful.Container
+	openapiv3Config    *common.OpenAPIV3Config
+	delegate           http.Handler
+
+	services         *lru.Cache
+	openAPIV3Service *handler3.OpenAPIService
+}
+
+func newOpenAPIHandler(
+	apiSetRetriever apidefinition.APIDefinitionSetGetter,
+	goRestfulContainer *restful.Container,
+	openapiv3Config *common.OpenAPIV3Config,
+	serviceCacheSize int,
+	delegate http.Handler) *openAPIHandler {
+	openAPIV3Service := handler3.NewOpenAPIService()
+
+	return &openAPIHandler{
+		apiSetRetriever:    apiSetRetriever,
+		goRestfulContainer: goRestfulContainer,
+		openapiv3Config:    openapiv3Config,
+		delegate:           delegate,
+
+		services:         lru.New(serviceCacheSize),
+		openAPIV3Service: openAPIV3Service,
+	}
+}
+
+func (o *openAPIHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	log := klog.FromContext(ctx).WithValues("path", req.URL.Path)
+	apiDomainKey := dyncamiccontext.APIDomainKeyFrom(ctx)
+
+	// Get all available APIDefinitions for this Virtual Workspace
+	apiSet, hasLocationKey, err := o.apiSetRetriever.GetAPIDefinitionSet(ctx, apiDomainKey)
+	if err != nil {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewInternalError(fmt.Errorf("unable to determine API definition set: %w", err)),
+			errorCodecs, schema.GroupVersion{},
+			w, req)
+		return
+	}
+	if !hasLocationKey {
+		o.delegate.ServeHTTP(w, req)
+		return
+	}
+
+	// Get the OpenAPI service from cache or create a new one
+	key, err := apiConfigurationKey(apiSet)
+	if err != nil {
+		responsewriters.InternalError(w, req, err)
+		return
+	}
+	log = log.WithValues("key", key)
+	entry, ok := o.services.Get(key)
+	if !ok {
+		log.V(7).Info("Generating OpenAPI v3 specs")
```

Comment:
this is a heavy operation, isn't it? Shouldn't we lock the saved key somehow and let other parties with the same interest wait? This even applies to one client collecting spec for example.

A common pattern for that: store an item struct in the lru cache with a wait channel:
```golang
type item struct {
  done <-chan struct{}
  err error
  service http.Handler
}
```
Every cache hit will wait for `done` before accessing `err` or `service`. The go routine owning this will add it to the lru cache before starting the heavy work. It will set `err` or `service` and then close `done` to wake all other waiters up.

Reply:
I implemented something based on that. I would very appreciate if you can take a look and let me know if it makes any sense.